servers/src/common/hooks.rs

```diff
@@ -0,0 +1,298 @@
+	/// The hyper client to be used for all requests
+	client: Client<HttpConnector>,
+	/// The tokio event loop
+	runtime: Runtime,
+}
+
+impl WebHook {
+	/// Instantiates a Webhook struct
+	pub fn new(
+		tx_received_url: Option<hyper::Uri>,
+		header_received_url: Option<hyper::Uri>,
+		block_received_url: Option<hyper::Uri>,
+		block_accepted_url: Option<hyper::Uri>,
+	) -> WebHook {
+		WebHook {
+			tx_received_url,
+			block_received_url,
+			header_received_url,
+			block_accepted_url,
+			client: Client::new(),
+			runtime: Runtime::new().unwrap(),
+		}
+	}
+
+	/// Instantiates a Webhook struct from a configuration file
+	pub fn from_config(config: &WebHooksConfig) -> WebHook {
+		WebHook::new(
+			parse_url(&config.tx_received_url),
+			parse_url(&config.header_received_url),
+			parse_url(&config.block_received_url),
+			parse_url(&config.block_accepted_url),
+		)
+	}
+
+	fn post(&self, url: hyper::Uri, data: String) {
+		let mut req = Request::new(Body::from(data));
+		*req.method_mut() = Method::POST;
+		*req.uri_mut() = url.clone();
+		req.headers_mut().insert(
+			hyper::header::CONTENT_TYPE,
+			HeaderValue::from_static("application/json"),
+		);
+
+		let future = self
+			.client
+			.request(req)
+			.map(|_res| {})
+			.map_err(move |_res| {
+				warn!("Error sending POST request to {}", url);
+			});
+
+		let handle = self.runtime.executor();
+		handle.spawn(future);
+	}
+}
+
+impl ChainEvents for WebHook {
+	fn on_block_accepted(&self, block: &core::Block, _status: &BlockStatus) {
+		match &self.block_accepted_url {
+			None => return,
+			Some(url) => {
+				let payload = match to_string(block) {
+					Ok(serialized) => serialized,
+					Err(_) => {
+						error!(
+							"Failed to serialize block {} at height {}",
+							block.hash(),
+							block.header.height
+						);
+						return;
+					}
+				};
+				self.post(url.clone(), payload);
+			}
+		}
+	}
+}
+
+impl NetEvents for WebHook {
+	/// Triggers when a new transaction arrives
+	fn on_transaction_received(&self, tx: &core::Transaction) {
+		match &self.tx_received_url {
+			None => return,
+			Some(url) => {
+				let payload = match to_string(tx) {
+					Ok(serialized) => serialized,
+					Err(_) => {
+						error!("Failed to serialize transaction {}", tx.hash());
+						return;
+					}
+				};
+				self.post(url.clone(), payload);
+			}
+		}
+	}
+
+	/// Triggers when a new block arrives
+	fn on_block_received(&self, block: &core::Block, _addr: &SocketAddr) {
+		match &self.block_received_url {
+			None => return,
+			Some(url) => {
+				let payload = match to_string(block) {
+					Ok(serialized) => serialized,
+					Err(_) => {
+						error!(
+							"Failed to serialize block {} at height {}",
+							block.hash(),
+							block.header.height
+						);
+						return;
+					}
+				};
+				self.post(url.clone(), payload);
+			}
+		}
+	}
+
+	/// Triggers when a new block header arrives
+	fn on_header_received(&self, header: &core::BlockHeader, _addr: &SocketAddr) {
```

Comment:
It seems we could make one generic function and call it from all handlers, wdyt?

Reply:
Yes makes sense

Reply:
fixed